luigi/contrib/beam_dataflow.py

```diff
@@ -0,0 +1,451 @@
+            output.append(f(self.dataflow_params.disk_size_gb, self.disk_size_gb))
+        if self.worker_machine_type:
+            output.append(f(self.dataflow_params.worker_machine_type, self.worker_machine_type))
+        if self.worker_disk_type:
+            output.append(f(self.dataflow_params.worker_disk_type, self.worker_disk_type))
+        if self.network:
+            output.append(f(self.dataflow_params.network, self.network))
+        if self.subnetwork:
+            output.append(f(self.dataflow_params.subnetwork, self.subnetwork))
+        if self.job_name:
+            output.append(f(self.dataflow_params.job_name, self.job_name))
+        if self.service_account:
+            output.append(f(self.dataflow_params.service_account, self.service_account))
+        if self.labels:
+            output.append(f(self.dataflow_params.labels, json.dumps(self.labels)))
+
+        return output
+
+    def _format_input_args(self):
+        job_input = self.input()
+        if isinstance(job_input, luigi.Target):
+            job_input = {"input": job_input}
+        elif not isinstance(job_input, dict):
+            raise ValueError("Input (requires()) must be dict type")
+
+        if not isinstance(self.file_pattern(), dict):
+            raise ValueError('file_pattern() must return a dict type')
+
+        input_args = []
+
+        for (name, targets) in job_input.items():
+            uri_targets = luigi.task.flatten(targets)
+            uris = [self._targets_to_uri_getter.get(
+                uri_target.__class__)(uri_target) for uri_target in uri_targets]
+            if isinstance(targets, dict):
+                """
+                If targets is a dict that means it had multiple outputs.
+                Make the input args in that case "<input key>-<task output key>"
+                """
+                names = ["%s-%s" % (name, key) for key in targets.keys()]
+            else:
+                names = [name] * len(uris)
+            for (arg_name, uri) in zip(names, uris):
+                pattern = self.file_pattern().get(name, 'part-*')
+                input_args.append(
+                    "--%s=%s" %
+                    (arg_name, uri.rstrip('/') + '/' + pattern)
+                )
+
+        return input_args
+
+    def _format_output_args(self):
+        job_output = self.output()
+        if isinstance(job_output, luigi.Target):
+            job_output = {"output": job_output}
+        elif not isinstance(job_output, dict):
+            raise ValueError("Input (requires()) must be dict type")
+
+        self.output_uris = {}
+        output_args = []
+
+        for (name, target) in job_output.items():
+            getter = self._targets_to_uri_getter.get(target.__class__)
+            uri = getter(target)
+
+            self.output_uris[name] = uri
+            output_args.append("--%s=%s" % (name, uri))
+
+        return output_args
+
+    @property
+    def _targets_to_uri_getter(self):
```

Comment:
I find this clever but obscure way to get the path of a Target.
I think that it will be simpler and more readable to write this function this way:
```python
  @property
  def _get_target_path(self, target):
    if isinstance(target, luigi.LocalTarget) or isinstance(target, gcs.GCSTarget):
      return target.path
    elif isinstance(target, bigquery.BigQueryTarget):
       "{}:{}.{}".format(target.project_id, target.dataset_id, target.table_id))
    else:
      raise ValueError("Target not supported")
```

Then when invoking this, lets say in line 406 you will write it like this:

```python
uris = [self._get_target_path(uri_target) for uri_target in uri_targets]
```

Reply:
hey, I agree that this method is a bit obscure as is. What I had in mind was that people might want to extend this method for custom Targets. For example, spotify's internal scio luigi extension supports some internally defined target types. can you think of a good way to extend the base target types here?

Reply:
Could users then just override `_get_target_path` ?

Also, should the name of a method that is expected to be overridden by subclasses should start with underscore?

Reply:
you're right, i was probably overthinking it. Updated it to be a static method `get_target_path` that can be overridden.